[WIP] Delay payout if fiat buyer has new account

ManfredKarrer opened the pull request:
This PR has merged https://github.com/bisq-network/bisq/pull/2753 as some of those features are used.

Implements proposal https://github.com/bisq-network/proposals/issues/77

Here are the relevant screens:

Create account:
<img width="640" alt="Screen Shot 2019-04-22 at 00 10 28" src="https://user-images.githubusercontent.com/1449498/56484239-31813700-6494-11e9-98f7-c13345516499.png">

Popup after account creation:
<img width="757" alt="Screen Shot 2019-04-22 at 00 11 10" src="https://user-images.githubusercontent.com/1449498/56484240-31813700-6494-11e9-9f9e-98ec1c2c5ff7.png">

Popup at create buy offer:
<img width="1185" alt="Screen Shot 2019-04-22 at 00 11 38" src="https://user-images.githubusercontent.com/1449498/56484241-31813700-6494-11e9-9dc5-dc2568656844.png">

Popup at take buy offer:
<img width="1128" alt="Screen Shot 2019-04-22 at 00 12 35" src="https://user-images.githubusercontent.com/1449498/56484242-31813700-6494-11e9-8f04-4ecd9671a60d.png">

Popup at confirm fiat receipt:
<img width="1129" alt="Screen Shot 2019-04-22 at 00 14 07" src="https://user-images.githubusercontent.com/1449498/56484244-3219cd80-6494-11e9-84f6-4e90d4736cfe.png">

Status while waiting until buyers account matures:
<img width="1118" alt="Screen Shot 2019-04-22 at 00 14 44" src="https://user-images.githubusercontent.com/1449498/56484245-3219cd80-6494-11e9-97a5-4bbb74029c42.png">

After buyers account has matured:
<img width="816" alt="Screen Shot 2019-04-22 at 00 16 36" src="https://user-images.githubusercontent.com/1449498/56484246-3219cd80-6494-11e9-8075-6cbf1bc6c10a.png">

ManfredKarrer committed:
Merge branch 'fix-trade-peers-account-age' into delay-payout-if-fait-buyer-has-new-account

ManfredKarrer committed:
Fix wrong display of ages with 1 at the end (e.g. 31 days)

31 days was displayed as 31 day.

ManfredKarrer committed:
Add min required buyers account age

Implements https://github.com/bisq-network/proposals/issues/77

If fiat trade the buyer need to have an min required account age,
otherwise the payout will be delayed.

ManfredKarrer committed:
Add warning popup to offer book

ManfredKarrer committed:
Remove rounded fiat amounts popup

I think the features is now well known and can be removed.
The account age check will add a nde popup so we would get more popup
hell if we leave that one...

ManfredKarrer committed:
Add popup to offer maker

If offer maker for buying BTC with Fiat has an account selected which is
less then the required account age for that method he gets a
popup displayed.

ManfredKarrer committed:
Refactor

ManfredKarrer committed:
Improve text

ManfredKarrer commented:
The handling of the trade period warnings and time display is still missing...

ManfredKarrer commented:
Another thing missing is an option for the seller as maker to accept or not accept younger accounts. We could add an option for min. account age for takers to an offer...

ManfredKarrer committed:
Handle trade period (WIP)

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -2644,7 +2660,10 @@ payment.limits.info=Please be aware that all bank transfers carry a certain amou
+  Please note that there are no limits on the total number of times you can trade.\n\n\
+  If you are the Bitcoin buyer the payout of the purchased Bitcoin will be delayed until you reach the required \
+  account age of {3}. You still can trade with a newer account but you will not receive the Bitcoin before \
+  you account age reaches {3}.
```
```
Please note that there are no limits on the number of times you can trade.\n\n
For trades where you are the bitcoin buyer, payout of the purchased bitcoin will be \
delayed until you reach the required account age of {3}. You are free to place more \
trades in the interim, but you will not receive the bitcoin for these trades before \
the account age reaches {3}.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -2292,8 +2305,10 @@ popup.shutDownInProgress.msg=Shutting down application can take a few seconds.\n
 
-popup.roundedFiatValues.headline=New privacy feature: Rounded fiat values
-popup.roundedFiatValues.msg=To increase privacy of your trade the {0} amount was rounded.\n\nDepending on the client version you''ll pay or receive either values with decimals or rounded ones.\n\nBoth values do comply from now on with the trade protocol.\n\nAlso be aware that BTC values are changed automatically to match the rounded fiat amount as close as possible.
+popup.immatureBuyerAccountAge.createOffer.msg=Your account has been created {0} ago.\n\n\
```
```
popup.immatureBuyerAccountAge.createOffer.msg=Your account was created {0} ago.\n\n\
To protect against bank chargeback fraud, this payment method requires a minimum account age of {1} for the bitcoin buyer.\n\n\
You can still create this offer, but the payout of the bitcoin will be delayed until your account age reaches {2}.\n\n\
This is a necessary protection against bank chargeback fraud.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -356,7 +356,10 @@ offerbook.warning.tradeLimitNotMatching=Your payment account has been created {0
+offerbook.warning.buyerHasImmatureAccount=The peer''s account age is {0} and therefor less than the required \
+  min. account age for Bitcoin buyers of {1}.\n\n\
+  You can still take that offer but the payout will be delayed until the buyer''s account age reaches {2}.
+offerbook.warning.buyerHasImmatureAccount.confirm=Ok, I understand and still want to take that offer.
```
```
offerbook.warning.buyerHasImmatureAccount.confirm=Ok, I understand and still want to take this offer.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -356,7 +356,10 @@ offerbook.warning.tradeLimitNotMatching=Your payment account has been created {0
   The min. trade amount of the offer is: {2}.\n\n\
   You cannot take that offer at the moment. Once your account is older than 2 months this restriction gets removed.
-
+offerbook.warning.buyerHasImmatureAccount=The peer''s account age is {0} and therefor less than the required \
```
```
offerbook.warning.buyerHasImmatureAccount=The peer''s account age is {0} and therefore less than the required \
minimum account age for bitcoin buyers of {1}.\n\n\
You can still take this offer, but the payout will be delayed until the buyer''s account age reaches {2}.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -654,6 +657,14 @@ portfolio.pending.step3_seller.onPaymentReceived.name=Please also verify that th
 portfolio.pending.step3_seller.onPaymentReceived.confirm.headline=Confirm that you have received the payment
 portfolio.pending.step3_seller.onPaymentReceived.confirm.yes=Yes, I have received the payment
+portfolio.pending.step3_seller.status.delayedPayoutInfo=Please open your Bisq application after {0} to release the Bitcoin.
+portfolio.pending.step3_seller.popup.delayedPayoutInfo=The buyer''s account has been created {0} ago.\n\n\
```
```
portfolio.pending.step3_seller.popup.delayedPayoutInfo=The buyer''s account was created {0} ago.\n\n\
To reduce risks with bank chargebacks, the payout of the bitcoin is delayed until {1}.\n\n\
Please be sure to open your Bisq application after that date to release the bitcoin to the buyer, and to get your security deposit refunded.\n\n\
In case you get a chargeback request from your bank, open a dispute with Cmd + o or Ctrl + o.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -654,6 +657,14 @@ portfolio.pending.step3_seller.onPaymentReceived.name=Please also verify that th
+In case you would get a chargeback request from your bank open a dispute with Cmd + o or Ctrl + o.
+
+portfolio.pending.step3_seller.button.release=Release Bitcoin
+portfolio.pending.step3_seller.status.confirmRelease=The buyer''s account is now {0} days old. Please confirm to release the Bitcoin.
```
```
portfolio.pending.step3_seller.status.confirmRelease=The buyer''s account is now {0} days old. Please confirm to release the bitcoin.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -654,6 +657,14 @@ portfolio.pending.step3_seller.onPaymentReceived.name=Please also verify that th
 portfolio.pending.step3_seller.onPaymentReceived.note=Please note, that as soon you have confirmed the receipt, the locked trade amount will be released to the BTC buyer and the security deposit will be refunded.
 portfolio.pending.step3_seller.onPaymentReceived.confirm.headline=Confirm that you have received the payment
 portfolio.pending.step3_seller.onPaymentReceived.confirm.yes=Yes, I have received the payment
+portfolio.pending.step3_seller.status.delayedPayoutInfo=Please open your Bisq application after {0} to release the Bitcoin.
```
```
portfolio.pending.step3_seller.status.delayedPayoutInfo=Please open your Bisq application after {0} to release the bitcoin.
```

m52go reviewed (commented):
Improved wording throughout.

Nitpick: I prefer [this convention](https://livebook.manning.com/#!/book/grokking-bitcoin/chapter-1/v-6/10) of using bitcoin for the currency and Bitcoin for the network.

ManfredKarrer committed:
Merge branch 'master' into delay-payout-if-fait-buyer-has-new-account

# Conflicts:
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java

ManfredKarrer committed:
Update strings

ManfredKarrer committed:
Refactoring: Move classed to new package

Pure technical refactoring, no functionality change.

ManfredKarrer committed:
Add Account score domain (WIP)

Move methods for delayed payout to AccountScore.

ManfredKarrer committed:
Refactor: Move methods to domain (WIP)

ManfredKarrer committed:
Move account age related code to AccountCreationAgeService

We use AccountScoreService as main class which constructs the
parameters based on several protection tools. Currently there is only
account age based protection.

ManfredKarrer committed:
Fix delay function, refactorings

ManfredKarrer committed:
Rename method

ManfredKarrer committed:
Add methods for getting the AccountScoreCategory

ManfredKarrer commented:
Replaced by https://github.com/bisq-network/bisq/pull/2787
